code/datums/components/ghostrole_on_revive.dm

```diff
@@ -0,0 +1,100 @@
+/// Proc ghosts to enter the body when it get's revived
+/datum/component/ghostrole_on_revive
+	/// If revived and no ghosts, just die again?
+	var/refuse_revival_if_failed
+	/// Callback for when the mob is revived and has their body occupied by a ghost
+	var/datum/callback/on_succesful_revive
+
+/datum/component/ghostrole_on_revive/Initialize(refuse_revival_if_failed, on_succesful_revive)
+	. = ..()
+
+	src.refuse_revival_if_failed = refuse_revival_if_failed
+	src.on_succesful_revive = on_succesful_revive
+
+	ADD_TRAIT(parent, TRAIT_GHOSTROLE_ON_REVIVE, REF(src)) //for adding an alternate examination
+
+	if(ismob(parent))
+		prepare_mob(parent)
+	else if(istype(parent, /obj/item/organ/internal/brain))
+		var/obj/item/organ/internal/brain/brein = parent
+		if(brein.owner)
+			prepare_mob(brein.owner)
+		else
+			prepare_brain(brein)
+	else
+		return COMPONENT_INCOMPATIBLE
+
+/// Give the appropriate signals, and watch for organ removal
+/datum/component/ghostrole_on_revive/proc/prepare_mob(mob/living/liver)
+	RegisterSignal(liver, COMSIG_LIVING_REVIVE, PROC_REF(on_revive))
+	ADD_TRAIT(liver, TRAIT_GHOSTROLE_ON_REVIVE, REF(src))
+	liver.med_hud_set_status()
+
+	if(iscarbon(liver))
+		var/mob/living/carbon/carbon = liver
+		var/obj/item/organ/brain = carbon.get_organ_by_type(/obj/item/organ/internal/brain)
+		if(brain)
+			RegisterSignal(brain, COMSIG_ORGAN_REMOVED, PROC_REF(on_remove))
+
+/datum/component/ghostrole_on_revive/proc/on_remove(obj/item/organ/brain, mob/living/old_owner)
+	SIGNAL_HANDLER
+
+	REMOVE_TRAIT(old_owner, TRAIT_GHOSTROLE_ON_REVIVE, REF(src))
+	prepare_brain(brain)
+
+/datum/component/ghostrole_on_revive/proc/prepare_brain(obj/item/organ/brein)
+	SIGNAL_HANDLER
+
+	RegisterSignal(brein, COMSIG_ORGAN_IMPLANTED, PROC_REF(prepare_mob_from_brain))
+	UnregisterSignal(brein, COMSIG_ORGAN_REMOVED)
+
+/datum/component/ghostrole_on_revive/proc/prepare_mob_from_brain(obj/item/organ/internal/brain/brein, mob/living/owner)
+	SIGNAL_HANDLER
+
+	UnregisterSignal(brein, COMSIG_ORGAN_IMPLANTED)
+	prepare_mob(owner)
+
+/datum/component/ghostrole_on_revive/proc/on_revive(mob/living/aliver)
+	SIGNAL_HANDLER
+
+	INVOKE_ASYNC(src, PROC_REF(poll_ghosts), aliver)
+
+/datum/component/ghostrole_on_revive/proc/poll_ghosts(mob/living/aliver)
+	var/soul_eyes
+	var/obj/item/bodypart/head
+	// adds soulful SOUL PENDING eyes to indicate what's happening to observers
+
+	var/mob/living/carbon/human/hewmon
+	if(ishuman(aliver))
+		hewmon = aliver
+		head = hewmon.get_bodypart(BODY_ZONE_HEAD)
+		soul_eyes = new /datum/bodypart_overlay/simple/soul_pending_eyes ()
+		head?.add_bodypart_overlay(soul_eyes)
+		hewmon.update_body_parts()
```

Comment:
if head is null, eyes overlay isn't disposed of. Only run this if head isn't null.